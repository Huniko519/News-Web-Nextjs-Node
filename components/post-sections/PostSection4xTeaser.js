import React from 'react';
import PropTypes from 'prop-types';
import PostTeaser from '../post-items/PostTeaser';
import PostSectionTitle from './PostSectionTitle';
import PostSectionWrapper from './PostSectionWrapper';
import { dedupePostList } from '../../utils/post';

/**
 * The PostSection4xTeaser component outputs 4 posts as PostTeaser components into the following
 * layouts.
 *
 * Desktop (min-width: 1024px):
 * ╔════════╤════════╤════════╤════════╗
 * ║        │        │        │        ║
 * ║ Teaser │ Teaser │ Teaser │ Teaser ║
 * ║        │        │        │        ║
 * ╚════════╧════════╧════════╧════════╝
 *
 * Tablet (min-width: 768px):
 * ╒════════╤════════╕
 * │        │        │
 * │ Teaser │ Teaser │
 * │        │        │
 * ├────────┼────────┤
 * │        │        │
 * │ Teaser │ Teaser │
 * │        │        │
 * ╘════════╧════════╛
 *
 * Mobile:
 * ╒════════╕
 * │        │
 * │ Teaser │
 * │        │
 * ├────────┤
 * │        │
 * │ Teaser │
 * │        │
 * ├────────┤
 * │        │
 * │ Teaser │
 * │        │
 * ├────────┤
 * │        │
 * │ Teaser │
 * │        │
 * ╘════════╛
 *
 *
 * @param {items: [...posts]} props
 * @see '../post-items/PostTeaser'
 */
const PostSection4xTeaser = (props) => {
  const {
    items, title, link, className,
  } = props;
  const p = dedupePostList(items, 4);

  return (
    p.length > 0
      ? (
        <PostSectionWrapper type="PostSection4xTeaser" className={className}>
          <PostSectionTitle title={title} link={link} />
          <div className="inews__post-section__body">
            {p.map((post) => <PostTeaser key={post.id} post={post} />)}
          </div>
        </PostSectionWrapper>
      )
      : null
  );
};

PostSection4xTeaser.propTypes = {
  items: PropTypes.arrayOf(PropTypes.object).isRequired,
  title: PropTypes.string,
  link: PropTypes.string,
  className: PropTypes.string,
};

PostSection4xTeaser.defaultProps = {
  title: '',
  link: '',
  className: '',
};

export default PostSection4xTeaser;
