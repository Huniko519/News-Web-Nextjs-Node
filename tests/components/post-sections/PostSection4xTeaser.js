import React from 'react';
import PostSection4xTeaser from '../../../components/post-sections/PostSection4xTeaser';

const items = [
  {
    id: 1,
    title: {
      rendered: 'title 1',
    },
    breadcrumbs: [
      { name: 'News' },
    ],
    link: 'https://inews-article.com/test-article-1',
  },
  {
    id: 2,
    title: {
      rendered: 'title 2',
    },
    breadcrumbs: [
      { name: 'News' },
    ],
    link: 'https://inews-article.com/test-article-2',
  },
  {
    id: 3,
    title: {
      rendered: 'title 3',
    },
    breadcrumbs: [
      { name: 'News' },
    ],
    link: 'https://inews-article.com/test-article-3',
  },
  {
    id: 4,
    title: {
      rendered: 'title 4',
    },
    breadcrumbs: [
      { name: 'News' },
    ],
    link: 'https://inews-article.com/test-article-4',
  },
];

jest.mock('../../../utils/post', () => ({
  __esModule: true,
  dedupePostList: jest.fn((dedupeItems, count = 10) => dedupeItems.slice(0, count)),
}));

beforeEach(() => {
  jest.clearAllMocks();
});

test('renders the snapshot', () => {
  const postSection4xTeaser = shallow(<PostSection4xTeaser items={items} title="PostSection title" link="https://inews.test/news" />);
  expect(postSection4xTeaser).toMatchSnapshot();
});

it('should output with the correct class and data-type on the root element', () => {
  const postSection4xTeaser = shallow(<PostSection4xTeaser items={items} />);
  const postSectionWrapper = postSection4xTeaser.find('PostSectionWrapper');
  expect(postSectionWrapper.length).toBe(1);
  expect(postSectionWrapper.prop('type')).toBe('PostSection4xTeaser');
});

it('should call the <PostSectionTitle /> component with the correct props', () => {
  const allProps = shallow(<PostSection4xTeaser items={items} title="PostSection title" link="https://inews.test/news" />);
  const justTitle = shallow(<PostSection4xTeaser items={items} title="PostSection title" />);
  const noProps = shallow(<PostSection4xTeaser items={items} />);

  const allPropsTitle = allProps.find('PostSectionTitle');
  expect(allPropsTitle.length).toBe(1);
  expect(allPropsTitle.prop('title')).toBe('PostSection title');
  expect(allPropsTitle.prop('link')).toBe('https://inews.test/news');

  const justTitleTitle = justTitle.find('PostSectionTitle');
  expect(justTitleTitle.length).toBe(1);
  expect(justTitleTitle.prop('title')).toBe('PostSection title');
  expect(justTitleTitle.prop('link')).toBe('');

  const noPropsTitle = noProps.find('PostSectionTitle');
  expect(noPropsTitle.length).toBe(1);
  expect(noPropsTitle.prop('title')).toBe('');
  expect(noPropsTitle.prop('link')).toBe('');
});

it('should call the <PostTeaser /> component four times with the correct props', () => {
  const postSection4xTeaser = shallow(<PostSection4xTeaser items={items} />);
  const PostTeasers = postSection4xTeaser.find('PostTeaser');
  expect(PostTeasers.length).toBe(4);

  expect(PostTeasers.at(0).prop('post')).toMatchObject(items[0]);
  expect(PostTeasers.at(1).prop('post')).toMatchObject(items[1]);
  expect(PostTeasers.at(2).prop('post')).toMatchObject(items[2]);
  expect(PostTeasers.at(3).prop('post')).toMatchObject(items[3]);

});

it('should call the <PostTeaser /> a upto four times if less than 4 items are passed to it', () => {
  const lessItems = [{ id: 1 }, { id: 2 }];
  const postSection4xTeaser = shallow(<PostSection4xTeaser items={lessItems} />);
  const PostTeasers = postSection4xTeaser.find('PostTeaser');
  expect(PostTeasers.length).toBe(2);
});

it('should call the <PostTeaser /> a maximum of four times regardless of how many items are passed to it', () => {
  const moreItems = [{ id: 1 }, { id: 2 }, { id: 3 }, { id: 4 }, { id: 5 }];
  const postSection4xTeaser = shallow(<PostSection4xTeaser items={moreItems} />);
  const PostTeasers = postSection4xTeaser.find('PostTeaser');
  expect(PostTeasers.length).toBe(4);
});

it('output null if items is an empty array', () => {
  const postSection4xTeaser = shallow(<PostSection4xTeaser items={[]} />);
  expect(postSection4xTeaser.getElement()).toBe(null);
});
