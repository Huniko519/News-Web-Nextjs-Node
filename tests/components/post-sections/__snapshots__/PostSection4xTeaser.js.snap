// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the snapshot 1`] = `
<PostSectionWrapper
  className=""
  teaserPosition=""
  type="PostSection4xTeaser"
>
  <PostSectionTitle
    link="https://inews.test/news"
    title="PostSection title"
  />
  <div
    className="inews__post-section__body"
  >
    <PostTeaser
      key="1"
      post={
        Object {
          "breadcrumbs": Array [
            Object {
              "name": "News",
            },
          ],
          "id": 1,
          "link": "https://inews-article.com/test-article-1",
          "title": Object {
            "rendered": "title 1",
          },
        }
      }
      showAuthor={false}
      showPubDate={false}
    />
    <PostTeaser
      key="2"
      post={
        Object {
          "breadcrumbs": Array [
            Object {
              "name": "News",
            },
          ],
          "id": 2,
          "link": "https://inews-article.com/test-article-2",
          "title": Object {
            "rendered": "title 2",
          },
        }
      }
      showAuthor={false}
      showPubDate={false}
    />
    <PostTeaser
      key="3"
      post={
        Object {
          "breadcrumbs": Array [
            Object {
              "name": "News",
            },
          ],
          "id": 3,
          "link": "https://inews-article.com/test-article-3",
          "title": Object {
            "rendered": "title 3",
          },
        }
      }
      showAuthor={false}
      showPubDate={false}
    />
    <PostTeaser
      key="4"
      post={
        Object {
          "breadcrumbs": Array [
            Object {
              "name": "News",
            },
          ],
          "id": 4,
          "link": "https://inews-article.com/test-article-4",
          "title": Object {
            "rendered": "title 4",
          },
        }
      }
      showAuthor={false}
      showPubDate={false}
    />
  </div>
</PostSectionWrapper>
`;
